core-js/src/main/javascript/cdf/components/BaseComponent.js

```diff
@@ -513,6 +524,90 @@ define([
       return opts || {};
     },
 
+    /**
+     * @summary Disposes this component, if it wasn't already disposed.
+     * @description Disposes this component, if it wasn't already disposed.
+     */
+    dispose: function() {
+      if(this.isDisposed) {
+        return;
+      }
+
+      this.isDisposed = true;
+
+      if(this.dashboard != null) {
+        this.dashboard.removeComponent(this);
+      } else {
+        // Usually, the whole dashboard is being disposed of.
+        this._releaseQuery();
+        this._releasePlaceholder();
+      }
+
+      this._disposeCore();
+    },
+
+    /**
+     * @summary Prepares the component for removal.
+     * @description Prepares the component for removal.
+     */
+    onWillRemove: function() {
+      if (this.dashboard && this.dashboard.refreshEngine) {
```

Comment:
I think it should be the dashboard doing this, don't you?

Reply:
Yes, it seals the functionality better.... the component isn't fully aware of the existence of a refresh engin.